### ISSUE TITLE
Auto-Grow Resources

### ISSUE THREAD

**Issue by tianmarin** [hashicorp/terraform#30177]
Terraform Version:
```
Terraform v1.1.0
```

Was checking the #27830 issues raised by @allcloud-jonathan.

In our team, we have been trying to understand how can we do something similar to what @allcloud-jonathan is proposing (equal_or_bigger_only lifecycle).
For example we have several SAP environment (which heavily rely on VMs and attached disks). If we automate the growth for the filesystems (let's say changing a disk from 100GB to 120GB), terraform will try to recreate it.

As stated by @apparentlymart `lifecycles` could prevent terraform destroy or even ignore size changing (`ignore_changes`). However we would loose the possibility to manage future changes using terraform (or even recreate it in case of disaster).

A great way to solve this issue would be to automate the disk growth by using terraform.
However it might be very complex to automate the complete process of alerting and reacting by updating terraform parameters, running the CI, getting a success or failure message to  then continue the workflow (for example, to execute some OS parameters to extend the filesystem). For that, we would have to create a complete set of parameters for each resource in other to the pipeline to produces de desired changes). However we try to be modular and try to use as little as parameter needed.

Other way would be to periodically execute the `terraform refresh` (`terraform apply -refresh-only -auto-approve`) command. Maybe this can import the new disk size, but a person would still be required to update the `*.tf` files to prevent destroying when other changes were in place.

---

What alternative would be better when trying to combine terraform management and "quick" non-terraform changes in the infrastructure?

Is there a way to automate this quick changes using terraform (the best example being disk growth automation based upon metrics and alarms)?

**Comment by apparentlymart** [comment]
Hi @tianmarin! Thanks for sharing this use-case, and context.

I think at the root of what you're describing here is an interesting (and non-straightforward) _separation of concerns_ architectural question. By which I mean: if the _overall system_ (Terraform Core, _and_ the relevant providers, _and_ the relevant remote systems all taken together) is expected to exhibit the behavior of being able to grow a disk as an in-place update without any disruption, what are all of the necessary steps, which component is responsible for each of them, and how would all of the components interact together to make sure the steps all happen in the correct sequence?

From what you've described here I see two steps that specifically _affect_ remote objects:
* At the virtual block storage level of abstraction, we must grow the reserved capacity of the volume, which basically just increases the number of blocks that can be stored in that volume without affecting the contents of any existing blocks. This operation would be initiated by the relevant Terraform provider against some platform-specific API, and I believe is already possible today in principle. (Whether existing providers _know_ that they can implement a size increase as an in-place update is a separate question, but that can at least be fixed by an addition to the provider within our existing architecture.)
* On top of the block storage layer there is a higher-level filesystem established by the operating system of the virtual machine that the volume is connected to, and it needs to be informed of the larger volume size so that it can update its own book-keeping to make use of the additional blocks now available, otherwise it will just ignore them and so they'd be wasted. Because the "filesystem" state is a mixture of data stored in the volume _and_ data stored in the RAM of the virtual machine it's connected to, making _this_ change requires running software inside the virtual machine, and in the current overall architecture there is no component which is clearly responsible for initiating that action in response to a Terraform configuration change.

To me, what makes this problem particularly tricky is that to solve it would effectively make Terraform act as the arbitrator between two systems which have a functional dependency between them but where they each aren't aware of that dependency: the block storage service knows nothing about the filesystem and the filesystem doesn't know about the block storage service specifically.

Normally in Terraform we resolve that sort of interaction by having a resource type representing each component and having the user declare a dependency relationship between two resources of those types. Hypothetically, something like this:

```hcl
resource "volume" "example" {
  size = 8
}

resource "filesystem" "example" {
  volume_id = volume.example.id
  size      = volume.example.size
}
```

In the hypothetical universe where this were possible, updating the size of `volume.example` to 16 would cause Terraform to see that the `filesystem.example` size had changed too, and so it would plan to first update-in-place the volume's size, and _then_ update-in-place the filesystem's size, due to the implied dependency between the two resources. That would achieve the correct order of operations to get the desired result, assuming that it were actually possible to perform both of those operations in-place without any disruption to other components.

That design isn't _obviously_ possible here because the "filesystem" is hidden from Terraform, encapsulated inside a virtual machine whose state is opaque to the cloud platform API and therefore in turn to Terraform.

One possible way to resolve that problem would be to run a special service inside the virtual machine which effectively provides a network API for the filesystem, and then a specialized Terraform provider could interact with that service to "update-in-place" the filesystem's settings. However, I don't know of any such service, and even if one did exist it seems counter-intuitive to run a network API for a Linux kernel data structure.

I don't have any ready answer to solve this. Technically you could _today_ write a special disk-management API and associated Terraform provider to implement the above, but that would be a lot of work in any situation where managing disks is not a crucial part of the product you're building; it would make more sense for a platform vendor to provide that feature to you, rather than for you to implement it yourself, but the platform vendors don't _currently_ seem interested in working at that level of abstraction.

**Comment by tianmarin** [comment]
Hi @apparentlymart,

First of all thank you for your great response. However I think an even more basic issue 😭 

Let's pretend I have solved the "orchestration" process:
- [x] Monitor the filesystems
- [x] Raise an alarm
- [x] Know which disk is related to what filesystem (even if its a logical volume LVM configuration )
- [ ] Call some API to grow the filesystem (might be Terraform or Cloud API)  👈🏻 
- [x] Expand the filesystem at OS configuration


For a very "simple"  scenario let's take the same example:
```tf
variable "volume_size"{
  default = 8
}
resource "volume" "example" {
  size = var.volume_size
}
```

I see two ways of executing the growth without human intervention:
- By cloud API and learn how to manage the drift 
- By directly executing the terraform pipeline changing the parameter.

Either way, executing the terraform pipeline would need the orchestrator to send the new value (to override each default value) , or terraform to extract the new value from somewhere (external data source?).

Maybe for one disk this could represent a challenge, but for hundreds of disks?

**Comment by apparentlymart** [comment]
Thanks for sharing that additional context, @tianmarin.

A part I hadn't understood on my first read is that you are intending to run Terraform as part of an automated response to detecting a low disk space condition, so thank you for sharing the additional information that made me notice that. Of course, how exactly you run Terraform is not something Terraform itself particularly "cares about", but it does indeed add the extra requirement of passing in an overridden volume size via a variable.

Terraform will see when it refreshes the "volume" in future that the size has increased, and will send that new value to the underlying provider. The provider would then be responsible for deciding how to make that change, and so I think the requirement you are asking for here seems to really just be a missing provider feature: as far as I can tell, there's nothing missing _in Terraform Core_ to meet this use-case, and instead it's more that the underlying cloud platform API (and, therefore, the Terraform provider wrapping that API) doesn't include a feature for growing a filesystem in an already-running system.

As I was describing in my previous comment, it seems like such a provider could exist _in theory_, with Terraform Core's current features. The challenge is that there's no existing API (that I know of) which can instruct a Linux system to resize a filesystem at runtime, and no Terraform provider for that hypothetical API.

---

Out of curiosity I did a little research into how resizing a filesystem to match an enlarged volume typically works on a Linux system. The clearest reference I found was [the Amazon EC2 documentation on growing a filesystem](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/recognize-expanded-volume-linux.html).

In particular, I notice that after increasing the EBS volume size the `lsblk` command seems to immediately start returning the newly-grown volume size, even though the partitions on that volume are still small enough to fit in the original capacity.

I don't know the internal details of how exactly Linux becomes aware of that larger volume size, but I have to assume that the EC2 virtual machine layer is generating some sort of configuration change event that the Linux kernel reacts to, adjusting its internal records of block device sizes.

That made me curious as to whether that event is exposed to userspace code via udev and after some research I found [a 2013 Linux patch to add a udev event for filesystem resize](https://lore.kernel.org/lkml/512B1637.5060107@redhat.com/T/), whose commit message includes some example udev rules to automatically resize the filesystem to fit the block device.

The exact configuration of your Linux systems is of course far outside of my scope as a Terraform maintainer, but if I were faced with the problem you described I would strongly consider using a mechanism like this so that the software running in these virtual machines could _automatically_ respond to a volume size increase, without any special orchestration from outside. That would then in principle allow you to use Terraform to resize the block device and then have that in turn _automatically_ grow the filesystem in your VM, without Terraform having any special awareness of it.

---

Given that Terraform Core already seems to have the necessary lifecycle to implement what you're describing (it's just that there's no current remote API for a Terraform provider to interact with to get the result you want), and given that there appears to be a feature integrated into Linux itself to avoid the need for this extra separate orchestration step altogether, my instinct is that there aren't any changes needed to Terraform Core to meet this use-case.

Given that, I'm going to close this issue just to reflect that there's nothing to be done _in this particular repository_ to meet your use-case. However, if you'd like to discuss further the possibility of implementing a custom provider to orchestrate the filesystem resize action then that could be a good topic for a topic in [our community forum](https://discuss.hashicorp.com/c/terraform-core/27).

The idea of configuring udev to automatically resize the event is not in Terraform's scope and so I don't think the Terraform community forum would be the best place to research _that_, but I hope the links I included above are useful to give you a starting point for further research if you want to pursue that further.

Thanks!
